### PR TITLE
school library: refactor

### PR DESCRIPTION
**Pull request by Zeph180**
1. Ensure YAGNI, KISS, and DRY rules and Ruby style guide are respected.
2. Follow SOLID principles.
3. Pay special attention to the "Single-responsibility Principle".
4. Make sure that the role of main.rb is only to start your code and ask the user for the initial choice of options (or to quit).
5. Extract some logic into a separate class(es).

Project requirements

At this point your codebase is pretty big. The main.rb file may be especially large at this point. Let's try to clean things up!

    Make sure YAGNI, KISS, and DRY rules and Ruby style guide are respected.
    Follow SOLID principles.
    Pay special attention to the "Single-responsibility Principle".
        Make sure that the role of main.rb is only to start your code and ask the user for the initial choice of options (or to quit).
        Extract some pieces of logic into separate class(es).

### CODE REVIEW

File: main_menu.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    '5' => :create_rental,
+    '6' => :list_rentals,
+    '7' => :exit
+  }.freeze
+
+  def initialize(app)
+    @app = app
+  end
+
+  # rubocop:disable Metrics/MethodLength
```
Comment (danmainah) [inline main_menu.rb]: - [ ] @Zeph180 and @djo1975 it's not good practice disabling the linter check in your code, I would advise you to try and fix the error, disabling the linter should be the last resort after trying and seeking help without success.

- [ ] @team I noted you still have the input and input validation in your app class  as shown by one example below, all inputs and input validation code need to be in the same class in order for you to achieve the `Single responsibility rule`, kindly ensure all requests for input and input validation should've been placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate the input(main_menu.rb).
image

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @Zeph180 and @djo1975 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

✔️ No linter errors
✔️ Gitflow was correctly used
✔️  Good design


# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: main_menu.rb (hunk 1 above)
Comment (Zeph180) [reply main_menu.rb]: @danmainah 
Thank you for this feedback.
However, it does not provide the entire issues to be fixed considering the fact that you said "One example below" Could you please throw more light on the other issues we are supposed to fix

Comment (danmainah) [reply main_menu.rb]: - [ ] @Zeph180 Any code that requests user input and input validation should be moved from the app.rb file.

Comment (djo1975) [reply main_menu.rb]: **At this stage we have to re-factorize main.rb**
Project requirements

At this point your codebase is pretty big. The main.rb file may be especially large at this point. Let's try to clean things up!

    Make sure YAGNI, KISS, and DRY rules and Ruby style guide are respected.
    Follow SOLID principles.
    Pay special attention to the "Single-responsibility Principle".
        Make sure that the role of main.rb is only to start your code and ask the user for the initial choice of options (or to quit).
        Extract some pieces of logic into separate class(es).

File: app.rb [hunk 2]
```diff
@@ -1,124 +1,110 @@
...
+# Create a person (teacher or student)
+def create_person(people)
+  puts 'Do you want to create a student(1) or a teacher(2)?[Input the number]: '
+  choice = gets.chomp.to_i
+  print 'Enter age: '
+  age = gets.chomp.to_i
+  print 'Enter name: '
+  name = gets.chomp
+  print 'Does the person have parent permission? (y/n): '
+  parent_permission = gets.chomp.downcase == 'y'
```
Comment (mikethreels) [inline app.rb]: - [x] as the previous CR mentioned any code in relation to input and input validation should be placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input. this includes all the `gets.chomp` parts of the code.

File: app.rb [hunk 3]
```diff
@@ -1,124 +1,110 @@
...
+end
 
-    book_id = gets.chomp.to_i
+def create_teacher(people, age, name, parent_permission)
+  print 'Enter specialization: '
+  specialization = gets.chomp
+  person = Teacher.new(people.length + 1, age, name, parent_permission, specialization)
+  people << person
+  puts "Created teacher #{person.name}"
+end
```
Comment (mikethreels) [inline app.rb]: - [x] as the previous CR mentioned any code in relation to input and input validation should be placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input. this includes all the `gets.chomp` parts of the code.

File: app.rb [hunk 4]
```diff
@@ -1,124 +1,110 @@
...
+# Create a book
+def create_book(books)
+  print 'Enter title: '
+  title = gets.chomp
+  print 'Enter author: '
+  author = gets.chomp
+  book = Book.new(title, author)
+  books << book
+  puts "Created book #{book.title} by #{book.author}"
+end
```
Comment (mikethreels) [inline app.rb]: - [x] as the previous CR mentioned any code in relation to input and input validation should be placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input. this includes all the `gets.chomp` parts of the code.

File: app.rb [hunk 5]
```diff
@@ -1,124 +1,110 @@
...
 
-    person_id = gets.chomp.to_i
+# Create a rental
+def create_rental(people, books, rentals)
+  puts 'Select a book from the following list by number:'
+  books.each { |book| puts "[#{book.id}] #{book.title} by #{book.author}" }
+  book_id = gets.chomp.to_i
+  book = books.find { |b| b.id == book_id }
```
Comment (mikethreels) [inline app.rb]: - [x] as the previous CR mentioned any code in relation to input and input validation should be placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input. this includes all the `gets.chomp` parts of the code.

File: app.rb [hunk 6]
```diff
@@ -1,124 +1,110 @@
...
+  puts "Created rental of '#{book.title}' by #{book.author} to #{person.name} on #{rental_date}"
+end
 
-    puts 'Rentals:'
-    rentals.each do |rental|
-      puts "Date: #{rental.date}, Book: '#{rental.book.title}' by #{rental.book.author}"
+def list_rentals(people, rentals)
+  puts 'Enter ID of person:'
+  person_id = gets.chomp.to_i
+  person = people.find { |p| p.id == person_id }
```
Comment (mikethreels) [inline app.rb]: - [x] as the previous CR mentioned any code in relation to input and input validation should be placed in the main.rb file and/or classes/methods whose responsibilities are to request or validate input. this includes all the `gets.chomp` parts of the code.

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi Team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- good job communicating with the previous CR.
## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mukhammadrizooff — APPROVED** [review]
Hi @TEAM, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
- Correct GitFlow ✔️ 
- No linter errors ✔️ 
- Good implementations ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by djo1975** [comment]
THX :)

**Pull request merged**